### PR TITLE
Update default email regex to allow special characters in the email username

### PR DESCRIPTION
**Pull request by ashanthamara**
### Proposed changes in this pull request
This PR will improve the BE regex validation logic for email usernames to allow special characters in the email.
`! # $ ' + \ = ^ _ . { | } ~ - &`

### CODE REVIEW

File: components/input-validation-mgt/org.wso2.carbon.identity.input.validation.mgt/src/main/java/org/wso2/carbon/identity/input/validation/mgt/model/handlers/UsernameValidationConfigurationHandler.java [hunk 1]
```diff
@@ -84,22 +85,20 @@ public ValidationConfiguration getDefaultValidationConfiguration(String tenantDo
...
-            } else {
-                rules.add(getRuleConfig(EmailFormatValidator.class.getSimpleName(),
+            if (!usernameRegEx.isEmpty() && !DEFAULT_EMAIL_JS_REGEX_PATTERN.equals(usernameRegEx)
+                    && isAlphaNumericValidationByDefault()) {
+                rules.add(getDefaultLengthValidatorRuleConfig());
+                rules.add(getRuleConfig(AlphanumericValidator.class.getSimpleName(),
                         ENABLE_VALIDATOR, Boolean.TRUE.toString()));
                 configuration.setRules(rules);
+            } else if (PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain().equals("carbon.super")) {
+                // Return the JSRegex if the tenant domain is carbon.super
```
Comment (kayathiri4) [inline components/input-validation-mgt/org.wso2.carbon.identity.input.validation.mgt/src/main/java/org/wso2/carbon/identity/input/validation/mgt/model/handlers/UsernameValidationConfigurationHandler.java:95]:
```suggestion
                // Return the JSRegex if the tenant domain is carbon.super.
```
And this preserves the existing behavior for Asgardeo users

**Review by kayathiri4 — APPROVED** [review]

File: components/input-validation-mgt/org.wso2.carbon.identity.input.validation.mgt/src/main/java/org/wso2/carbon/identity/input/validation/mgt/model/handlers/UsernameValidationConfigurationHandler.java [hunk 2]
```diff
@@ -84,22 +85,20 @@ public ValidationConfiguration getDefaultValidationConfiguration(String tenantDo
...
-                }
-            } else {
-                rules.add(getRuleConfig(EmailFormatValidator.class.getSimpleName(),
+            if (!usernameRegEx.isEmpty() && !DEFAULT_EMAIL_JS_REGEX_PATTERN.equals(usernameRegEx)
+                    && isAlphaNumericValidationByDefault()) {
+                rules.add(getDefaultLengthValidatorRuleConfig());
+                rules.add(getRuleConfig(AlphanumericValidator.class.getSimpleName(),
                         ENABLE_VALIDATOR, Boolean.TRUE.toString()));
                 configuration.setRules(rules);
+            } else if (PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain().equals("carbon.super")) {
```
Comment (Avarjana) [inline components/input-validation-mgt/org.wso2.carbon.identity.input.validation.mgt/src/main/java/org/wso2/carbon/identity/input/validation/mgt/model/handlers/UsernameValidationConfigurationHandler.java:94]: Shall we use a constant for carbon.super? There maybe existing one

File: features/claim-mgt/org.wso2.carbon.claim.mgt.server.feature/resources/conf/claim-config.xml [hunk 3]
```diff
@@ -67,7 +67,7 @@
 				<Required />
 				<DisplayOrder>6</DisplayOrder>
 				<SupportedByDefault />
-                <RegEx>^([a-zA-Z0-9_\+\.\-])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,10})+$</RegEx>
+                <RegEx>^([a-zA-Z0-9!#$'+\\=^_.{|}~\-&amp;])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,10})+$</RegEx>
```
Comment (Avarjana) [inline features/claim-mgt/org.wso2.carbon.claim.mgt.server.feature/resources/conf/claim-config.xml:70]: Don't we have to escape +,- and . ?

File: components/input-validation-mgt/org.wso2.carbon.identity.input.validation.mgt/src/main/java/org/wso2/carbon/identity/input/validation/mgt/model/handlers/UsernameValidationConfigurationHandler.java (hunk 2 above)
Comment (ashanthamara) [reply components/input-validation-mgt/org.wso2.carbon.identity.input.validation.mgt/src/main/java/org/wso2/carbon/identity/input/validation/mgt/model/handlers/UsernameValidationConfigurationHandler.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5568/commits/8a00dce0db28649472b6fd4777f9c02f893110a8

File: features/claim-mgt/org.wso2.carbon.claim.mgt.server.feature/resources/conf/claim-config.xml (hunk 3 above)
Comment (ashanthamara) [reply features/claim-mgt/org.wso2.carbon.claim.mgt.server.feature/resources/conf/claim-config.xml]: We don't need to escape characters inside [ ]. But adding escape for more safety.
Fixed with https://github.com/wso2/carbon-identity-framework/pull/5568/commits/8a00dce0db28649472b6fd4777f9c02f893110a8

**Review by kayathiri4 — APPROVED** [review]

**Pull request merged**